c/shared/source/uforead/uforead.c

```diff
@@ -1819,9 +1818,153 @@ static void reallocFDArray(ufoCtx h){
     h->top.FDArray.array = newFDArray;
 }
 
+static void parseXMLFile(ufoCtx h, char* filename, const char* filetype){
+    xmlDocPtr doc;
+    xmlNodePtr cur;
+    unsigned char* keyID;
+    char* keyName;
+
+    xmlKeepBlanksDefault(0);
+
+    h->cb.stm.xml_read(&h->cb.stm, h->stm.src, &h->src.buf, &doc);
+
+    if (doc == NULL) {
+        fatal(h, ufoErrParse, "Failed to parse '%s'.\n", "file");
+    }
+
+    cur = xmlDocGetRootElement(doc);
+    if (cur == NULL) {  // if document empty
+        xmlFreeDoc(doc);
+//        return(NULL);
+    }
+
+    if (!xmlStrEqual((cur)->name, (const xmlChar *) filetype)) {
+        fprintf(stderr, "document of the wrong type, root node != %s", filetype);
+        xmlFreeDoc(doc);
+//        return(NULL);
+    }
+
+    cur = (cur)->xmlChildrenNode;
+    while (cur && xmlIsBlankNode(cur)) {
+        cur = (cur) -> next;
+    }
+    if ( cur == 0 ) {
+        xmlFreeDoc(doc);
+//        return ( NULL );
+    }
+    if ((!xmlStrEqual((cur)->name, (const xmlChar *) "dict"))) {
+        fprintf(stderr, "Error reading outermost <dict> in %s.\n", "file");
+        xmlFreeDoc(doc);
+//        return(NULL);
+    }
+
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        keyID = NULL;
+        if ((xmlStrEqual(cur->name, (const xmlChar *) "key"))) {
+            parseKeyName(&keyID, cur);
+            if (keyID == NULL){
+                message(h, "Warning: Encountered empty <key></key>.");
+                cur = cur->next;
+                continue;
+            }
+            keyName = (char*) keyID;
+            cur = cur->next;
+            setFontDictKey(h, keyName, cur);
+        }
+        cur = cur->next;
+    }
+}
+
+static void parseKeyName(unsigned char* *keyID, xmlNodePtr cur){
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        if (!xmlStrEqual(cur->name, (const xmlChar *) "text"))
+            continue;
+        *keyID = xmlNodeGetContent(cur);
```

Comment:
If we want to use `keyID` as a char, maybe just cast it here?

Also, should we just `return` here? Two "text" nodes is unlikely but if it happens we'll probably leak the first assignment.